docs/configuration.md
# Configuring Dispatch

This document describes additional configuration options available to the Dispatch server itself.

## First Install

During a new install, you will be prompted first for a walkthough of the Installation Wizard. This wizard will help you get a few essential configuration options taken care of before begining.

Dispatch uses the same configuration system as [Starlette](https://www.starlette.io/config/).

By default, the config will be read from environment variables and/or `.env` files.

### General

#### `LOG_LEVEL` [optional][default: 'warning']

Controls the level of logging the application will perform during operations.

#### `STATIC_DIR` [optional][default: './src/static/dispatch/dist']

Controls where static content for the Dispatch UI should be served from. This can also be explicity set to `''` if you wish to serve static content outside of the Dispatch server.

#### `METRIC_PROVIDERS` [optional][default: ""]

A comma seperated list of metric providers Dispatch will send key system metrics to.

#### `SENTRY_DSN` [optional][default: none]

Optional configuration for using Sentry to report Dispatch errors.

#### `DISPATCH_HELP_EMAIL` [required]

Email address to be used by Dispatch when a help message is created.

#### `DISPATCH_HELP_SLACK_CHANNEL` [required]

Slack channel name to be used by Dispatch when a help message is created.

### Authentication

#### `JWKS_URL` [required]

Used by Dispatch's authentication backend to pull the JWKS public key from our specified provider. The result of this url is cached for up to 1 hour.

#### `VUE_APP_DISPATCH_OPEN_ID_CONNECT` [required]

Used by the Dispatch UI send the user via PKCE to a correct open id connect endpoint.

#### `VUE_APP_DISPATCH_CLIENT_ID` [required]

The client id to send to the open id connect endpoint.

### Persistence

#### `DATABASE_HOSTNAME` [required]

Dispatch relies on a `Postgres` database. This host name should point to a supporter version of `postgres`.

#### `DATABASE_CREDENTIALS` [required]

Credentials specified in `username:password` format to be used to authenticate to the `postgres` database.

#### `DATABASE_NAME` [optional][default: 'dispatch']

Allows the user to specify the database name for the `Dispatch` backend.

#### `DATABASE_PORT` [optional][default: '5432']

Allows the user to specify the database port fo the `Dispatch` backend.

### Models

### Incident Cost

#### `ANNUAL_COST_EMPLOYEE` [optional][default: '650000]

Used for incident cost modeling, specifices the total `all-in` cost for a average employee working on incidents.


docs/changelog.md
# Changelog

## 0.1.0 - 2042-12-03

### Added

* Initial release





src/dispatch/config.py
import logging
import os
import base64

from starlette.config import Config
from starlette.datastructures import CommaSeparatedStrings

# if we have metatron available to us, lets use it to decrypt our secrets in memory
try:
    import metatron.decrypt

    class Secret:
        """
        Holds a string value that should not be revealed in tracebacks etc.
        You should cast the value to `str` at the point it is required.
        """

        def __init__(self, value: str):
            self._value = value
            self._decrypted_value = (
                metatron.decrypt.MetatronDecryptor()
                .decryptBytes(base64.b64decode(self._value))
                .decode("utf-8")
            )

        def __repr__(self) -> str:
            class_name = self.__class__.__name__
            return f"{class_name}('**********')"

        def __str__(self) -> str:
            return self._decrypted_value


except Exception:
    from starlette.datastructures import Secret


config = Config(".env")

LOG_LEVEL = config("LOG_LEVEL", default=logging.WARNING)

DISPATCH_HELP_EMAIL = config("DISPATCH_HELP_EMAIL")
DISPATCH_HELP_SLACK_CHANNEL = config("DISPATCH_HELP_SLACK_CHANNEL")

# authentication
JWKS_URL = config("JWKS_URL")

DEFAULT_STATIC_DIR = os.path.join(
    os.path.abspath(os.path.dirname(__file__)), "static/dispatch/dist"
)
STATIC_DIR = config("STATIC_DIR", default=DEFAULT_STATIC_DIR)

# metrics
METRIC_PROVIDERS = config("METRIC_PROVIDERS", cast=CommaSeparatedStrings, default="")

# sentry middleware
SENTRY_DSN = config("SENTRY_DSN", cast=Secret, default=None)

# database
DATABASE_HOSTNAME = config("DATABASE_HOSTNAME")
DATABASE_CREDENTIALS = config("DATABASE_CREDENTIALS", cast=Secret)
DATABASE_NAME = config("DATABASE_NAME", default="dispatch")
DATABASE_PORT = config("DATABASE_PORT", default="5432")
SQLALCHEMY_DATABASE_URI = f"postgresql+psycopg2://{DATABASE_CREDENTIALS}@{DATABASE_HOSTNAME}:{DATABASE_PORT}/{DATABASE_NAME}"

# incident plugins
INCIDENT_PLUGIN_CONTACT_SLUG = config("INCIDENT_PLUGIN_CONTACT_SLUG", default="pandora-contact")
INCIDENT_PLUGIN_CONVERSATION_SLUG = config(
    "INCIDENT_PLUGIN_CONVERSATION_SLUG", default="slack-conversation"
)
INCIDENT_PLUGIN_DOCUMENT_SLUG = config(
    "INCIDENT_PLUGIN_DOCUMENT_SLUG", default="google-docs-document"
)
INCIDENT_PLUGIN_DOCUMENT_RESOLVER_SLUG = config(
    "INCIDENT_PLUGIN_DOCUMENT_RESOLVER_SLUG", default="dispatch-document-resolver"
)
INCIDENT_PLUGIN_EMAIL_SLUG = config(
    "INCIDENT_PLUGIN_EMAIL_SLUG", default="google-gmail-conversation"
)
INCIDENT_PLUGIN_GROUP_SLUG = config(
    "INCIDENT_PLUGIN_GROUP_SLUG", default="google-group-participant-group"
)
INCIDENT_PLUGIN_PARTICIPANT_SLUG = config(
    "INCIDENT_PLUGIN_PARTICIPANT_SLUG", default="dispatch-participants"
)
INCIDENT_PLUGIN_STORAGE_SLUG = config(
    "INCIDENT_PLUGIN_STORAGE_SLUG", default="google-drive-storage"
)
INCIDENT_PLUGIN_TICKET_SLUG = config("INCIDENT_PLUGIN_TICKET_SLUG", default="jira-ticket")
INCIDENT_PLUGIN_TASK_SLUG = config("INCIDENT_PLUGIN_TASK_SLUG", default="google-drive-task")

# incident resources
INCIDENT_STORAGE_ARCHIVAL_FOLDER_ID = config("INCIDENT_STORAGE_ARCHIVAL_FOLDER_ID")
INCIDENT_STORAGE_INCIDENT_REVIEW_FILE_ID = config("INCIDENT_STORAGE_INCIDENT_REVIEW_FILE_ID")
INCIDENT_DOCUMENT_INVESTIGATION_SHEET_ID = config("INCIDENT_DOCUMENT_INVESTIGATION_SHEET_ID")
INCIDENT_FAQ_DOCUMENT_ID = config("INCIDENT_FAQ_DOCUMENT_ID")

INCIDENT_NOTIFICATION_CONVERSATIONS = config(
    "INCIDENT_NOTIFICATION_CONVERSATIONS", cast=CommaSeparatedStrings, default=""
)

INCIDENT_NOTIFICATION_DISTRIBUTION_LISTS = config(
    "INCIDENT_NOTIFICATION_DISTRIBUTION_LISTS", cast=CommaSeparatedStrings, default=""
)

INCIDENT_DAILY_SUMMARY_ONCALL_SERVICE_ID = config(
    "INCIDENT_DAILY_SUMMARY_ONCALL_SERVICE_ID", default=None
)

INCIDENT_RESOURCE_TACTICAL_GROUP = config(
    "INCIDENT_RESOURCE_TACTICAL_GROUP", default="google-group-participant-tactical-group"
)
INCIDENT_RESOURCE_NOTIFICATIONS_GROUP = config(
    "INCIDENT_RESOURCE_NOTIFICATIONS_GROUP", default="google-group-participant-notifications-group"
)
INCIDENT_RESOURCE_FAQ_DOCUMENT = config(
    "INCIDENT_RESOURCE_FAQ_DOCUMENT", default="google-docs-faq-document"
)
INCIDENT_RESOURCE_INVESTIGATION_DOCUMENT = config(
    "INCIDENT_RESOURCE_INVESTIGATION_DOCUMENT", default="google-docs-investigation-document"
)
INCIDENT_RESOURCE_INVESTIGATION_SHEET = config(
    "INCIDENT_RESOURCE_INVESTIGATION_SHEET", default="google-docs-investigation-sheet"
)
INCIDENT_RESOURCE_INCIDENT_REVIEW_DOCUMENT = config(
    "INCIDENT_RESOURCE_INCIDENT_REVIEW_DOCUMENT", default="google-docs-incident-review-document",
)
INCIDENT_RESOURCE_INCIDENT_TASK = config(
    "INCIDENT_RESOURCE_INCIDENT_TASK", default="google-docs-incident-task"
)

# Incident Cost Configuration
ANNUAL_COST_EMPLOYEE = config("ANNUAL_COST_EMPLOYEE", cast=int, default="650000")
BUSINESS_HOURS_YEAR = config("BUSINESS_HOURS_YEAR", cast=int, default="2080")
